h2o-genmodel/src/main/java/hex/genmodel/algos/tree/SharedTreeMojoModel.java

```diff
@@ -177,16 +186,90 @@ public static double scoreTree(byte[] tree, double[] row, int nclasses, boolean
         }
     }
 
-    public static String getDecisionPath(double leafAssignment) {
+    public interface DecisionPathTracker<T> {
+        boolean go(int depth, boolean right);
+        T terminate();
+    }
+
+    public static class StringDecisionPathTracker implements DecisionPathTracker<String> {
+        private final char[] _sb = new char[64];
+        private int _pos = 0;
+        @Override
+        public boolean go(int depth, boolean right) {
+            _sb[depth] = right ? 'R' : 'L';
+            if (right) _pos = depth;
+            return true;
+        }
+        @Override
+        public String terminate() {
+            String path = new String(_sb, 0, _pos);
+            _pos = 0;
+            return path;
+        }
+    }
+
+    public static class LeafDecisionPathTracker implements DecisionPathTracker<LeafDecisionPathTracker> {
+        private final AuxInfoLightReader _auxInfo;
+        private boolean _wentRight = false; // Was the last step _right_?
+
+        // OUT
+        private int _nodeId = 0; // Returned when the tree is empty (consistent with SharedTreeNode of an empty tree)
+
+        private LeafDecisionPathTracker(byte[] auxTree) {
+          _auxInfo = new AuxInfoLightReader(new ByteBufferWrapper(auxTree));
+        }
+
+        @Override
+        public boolean go(int depth, boolean right) {
+          if (!_auxInfo.hasNext()) {
+            assert _wentRight || depth == 0; // this can only happen if previous step was _right_ or the tree has no nodes
```

Comment:
Terminal nodes can be both left or right. Left nodes know they are the terminal ones because of how the trees are represented (number of nodes in the left sub-tree). When you go in the "right" direction you don't know how big the right sub-tree is. Last step is always right (that is how the path is encoded in the Double value). That is why you know that if you run out of nodes the previous step had to be right (in the left node you would know you are in a terminal step).